### PR TITLE
FACES-3183 Create Cache and CacheFactory utilities

### PR DESCRIPTION
**Pull request by stiemannkj1**

### CODE REVIEW

File: alloy/src/main/java/com/liferay/faces/alloy/i18n/internal/I18nAlloyImpl.java [hunk 1]
```diff
@@ -35,4 +37,18 @@ public I18nAlloyImpl(I18n i18n) {
 	public String getBundleKey() {
 		return "i18n-alloy";
 	}
+
+	@Override
+	protected Integer getMaxCacheCapacity(ExternalContext externalContext) {
+
+		Integer maxCacheCapacity = null;
+		String maxCacheCapacityString = externalContext.getInitParameter(
+				"com.liferay.faces.alloy.i18n.AlloyI18nBundle.maxCacheCapacity");
```
Comment (ngriffin7a) [inline alloy/src/main/java/com/liferay/faces/alloy/i18n/internal/I18nAlloyImpl.java:46]: @stiemannkj1 Please use AlloyWebConfigParam instead for init params.

File: alloy/src/main/java/com/liferay/faces/alloy/taghandler/LoadConstants.java [hunk 2]
```diff
@@ -90,37 +91,70 @@ public LoadConstants(TagConfig config) throws Exception {
...
+			synchronized (LoadConstants.class) {
+
+				constantCache = LoadConstants.constantCache;
+
+				if (constantCache == null) {
+
+					FacesContext facesContext = facletContext.getFacesContext();
+					ExternalContext externalContext = facesContext.getExternalContext();
+					String maxCacheCapacityString = externalContext.getInitParameter(LoadConstants.class.getName() +
+							".maxCacheCapacity");
```
Comment (ngriffin7a) [inline alloy/src/main/java/com/liferay/faces/alloy/taghandler/LoadConstants.java:111]: @stiemannkj1 Also here, please use AlloyWebConfigParam instead for init params.

File: alloy/src/main/java/com/liferay/faces/alloy/taghandler/LoadConstants.java [hunk 3]
```diff
@@ -90,37 +91,70 @@ public LoadConstants(TagConfig config) throws Exception {
...
 	@Override
-	public void apply(FaceletContext facesContext, UIComponent parentUIComponent) throws IOException {
+	public void apply(FaceletContext facletContext, UIComponent parentUIComponent) throws IOException {
+
+		Cache<String, Map<String, Object>> constantCache = LoadConstants.constantCache;
+
+		// First check without locking (not yet thread-safe)
+		if (constantCache == null) {
+
+			// Second check with locking (thread-safe)
```
Comment (ngriffin7a) [inline alloy/src/main/java/com/liferay/faces/alloy/taghandler/LoadConstants.java:101]: @stiemannkj1 Please move this comment "Second check with locking (thread-safe)" to just above `if (constantCache == null) {` in order to stay consistent with usage of the double-check idom elsewhere in the codebase. For example: https://github.com/liferay/liferay-faces-bridge-api/blob/master/bridge-api/src/main/java/javax/portlet/faces/GenericFacesPortlet.java#L795

File: alloy/src/main/java/com/liferay/faces/alloy/taghandler/LoadConstants.java [hunk 4]
```diff
@@ -90,37 +91,70 @@ public LoadConstants(TagConfig config) throws Exception {
...
+			synchronized (LoadConstants.class) {
```
Comment (ngriffin7a) [inline alloy/src/main/java/com/liferay/faces/alloy/taghandler/LoadConstants.java:102]: @stiemannkj1 Are you certain that a TagHandler is not application scoped?

File: alloy/src/main/java/com/liferay/faces/alloy/taghandler/LoadConstants.java [hunk 5]
```diff
@@ -90,37 +91,70 @@ public LoadConstants(TagConfig config) throws Exception {
...
+
 				Class<?> clazz = Class.forName(classType);
 				Field[] fields = clazz.getFields();
 
 				for (Field field : fields) {
+
 					int modifiers = field.getModifiers();
 
-					if (Modifier.isStatic(modifiers)) {
+					if (Modifier.isPublic(modifiers) && Modifier.isStatic(modifiers) && Modifier.isFinal(modifiers)) {
```
Comment (ngriffin7a) [inline alloy/src/main/java/com/liferay/faces/alloy/taghandler/LoadConstants.java:146]: @stiemannkj1 Restricting to `public` should be OK, but restricting to `final` could cause a regression in downstream developer usage of `alloy:loadConstants` if they neglected to add the keyword in their code.

**Comment by ngriffin7a** [comment]
bp to 2.x and partially to 1.x see fixed typo: https://github.com/liferay/liferay-faces-alloy/commit/467c183b03cb26e32b5ec6c795b4284a06fbcc1d <br><br>:octocat: *Sent from [GH](http://nodegh.io).*

**Comment by ngriffin7a** [comment]
Merged, thank you. <br><br>:octocat: *Sent from [GH](http://nodegh.io).*

File: alloy/src/main/java/com/liferay/faces/alloy/taghandler/LoadConstants.java (hunk 5 above)
Comment (stiemannkj1) [reply alloy/src/main/java/com/liferay/faces/alloy/taghandler/LoadConstants.java:146]: @ngriffin7a, well that's just a bug in our code that people were exploiting. Constants must be `public static final` (at least as we have defined them although others have as well I think).

Comment (ngriffin7a) [reply alloy/src/main/java/com/liferay/faces/alloy/taghandler/LoadConstants.java:146]: @stiemannkj1 I agree that the concept of a constant should include finality, I also agree that we should have included the check for final from the beginning, but we need to guard against regressions downstream.

Comment (stiemannkj1) [reply alloy/src/main/java/com/liferay/faces/alloy/taghandler/LoadConstants.java:146]: @ngriffin7a, okay, adding a warning if field is non-`final` instead.

File: alloy/src/main/java/com/liferay/faces/alloy/taghandler/LoadConstants.java (hunk 4 above)
Comment (stiemannkj1) [reply alloy/src/main/java/com/liferay/faces/alloy/taghandler/LoadConstants.java:102]: @ngriffin7a, they are created when you hit the view that they are on. 1 is created per `alloy:loadConstant` on the page.

Comment (ngriffin7a) [reply alloy/src/main/java/com/liferay/faces/alloy/taghandler/LoadConstants.java:102]: @stiemannkj1 OK, good call using `synchronized (LoadConstants.class)` then.

**Pull request closed without merging**